ssm/observations.py

```diff
@@ -877,6 +877,34 @@ def sample_x(self, z, xhist, input=None, tag=None, with_noise=True):
             S = np.linalg.cholesky(self.Sigmas[z]) if with_noise else 0
             return mu + np.dot(S, npr.randn(D))
 
+    def hessian_expected_log_dynamics_prob(self, Ez, data, input, mask, tag=None):
+        assert np.all(mask), "Cannot compute Hessian of autoregressive obsevations with missing data."
+        assert self.lags == 1, "Does not compute Hessian of autoregressive observations with lags > 1"
+        T = data.shape[0]
+        K = self.K
+        D = self.D
+
+        # Ez is (TxK) array of expected discrete states
+
+        # diagonal blocks, size ((T, D, D))
+        inv_Sigmas = np.linalg.inv(self.Sigmas) # get inverse covariance matrices
+        dynamics_terms = np.array([A.T@inv_Sigma@A for A, inv_Sigma in zip(self.As, inv_Sigmas)]) # A^T Qinv A terms
+
+        # first part of diagonal blocks are inverse covariance matrices
+        diagonal_blocks = -1.0 * np.sum(Ez[:,:,None,None] * inv_Sigmas[None,:], axis=1)
```

Comment:
Why is this negated? I think the only negatives are on the off-diagonal blocks.

Reply:
We decided to represent the Hessian and negate downstream.